### PR TITLE
DEV-2213: Allow blob image URLs in CSP

### PR DESCRIPTION
**Pull request by nrh-cklimas**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

n/a

#### What's this PR do?

1. Allows `blob` image URLs in CSP.
2. Adds a new task, `make run-hybrid`, which locally serves the SPA through Django, parsing template tags in HTML and applying CSP policies.

#### Why are we doing this? How does it help us?

1. Gets us closer to being able to turn CSP on in prod.
2. The `make run-hybrid` task removes the template crud we've been seeing locally and makes it possible to debug CSP issues without using a review app.

#### How should this be manually tested? Please include detailed step-by-step instructions.

1. Edit a contribution page.
2. Choose a new image in the Setup tab. Confirm that the preview shows the new image you've selected.
3. Cancel out of changes and reload the page.
4. Remove an existing image in the Setup tab, then choose a new image. Confirm that the preview is also correct in this case.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

No. The CSP change is config and the stuff I added on the Django side to make `make run-hybrid` work is intended to only be active when doing local dev work.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

I'm not sure at all if the Django changes are idiomatic or a good idea. Also, I needed to put a manual check for whether pytest was running in view.py; otherwise it would try to proxy to a nonexistent SPA during pytest. Changing behavior based on whether tests are running doesn't seem great.

#### Has this been documented? If so, where?

I updated the README to describe the new task.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2213

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,6 +265,14 @@ make run-dev
 
 The React app will be available at `https://localhost:3000/`, and the Django admin will be available at `http://localhost:8000/nrhadmin/`.
 
+Alternately:
+
+```sh
+make run-hybrid
+```
+
+This reverses control locally. Django will serve the React app at `http://localhost:8000`, parsing template tags in the React app's HTML and enforcing CSP. If you use this task, you must set the environment variable `ENFORCE_CSP` to `false` or else the app won't load; the local app's `<script>` and `<style>` tags will not have the correct nonce attribute applied as they normally would when served in production.
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply README.md:274]: Why won't the nonces be correct? 

To be clear, this is a great improvement, and I'm happy to see the template crud going away. We're just so close to being able fully test CSP locally (inclusive of enforcing), I thought I'd ask.

File: revengine/urls.py [hunk 2]
```diff
@@ -55,12 +55,9 @@
 
 if settings.DEBUG:  # pragma: no cover Is covered, but reimport confuses coverage
     from django.conf.urls.static import static
-    from django.contrib.staticfiles.urls import staticfiles_urlpatterns
 
     import debug_toolbar
 
-    # Serve static and media files from development server
-    urlpatterns += staticfiles_urlpatterns()
```
Comment (benjaminEwhite) [inline revengine/urls.py:63]: Not sure if this is related or not, but _some_ images are breaking when running with `make run-hybrid`. When I go to login page for dashboard with `make run-hybrid` running, there are two broken images.


Screenshot 2023-01-12 at 10 08 31 AM


However, when I run using usual approach (`python manage.py runserver` in one tab and `npm run start:subdomains` in another), the images render.

**Review by benjaminEwhite — CHANGES_REQUESTED** [review]
@nrh-cklimas code change to solve for acute problem (image not appearing in preview) looks fine, and confirmed it works locally.

I love that we're going to be able to remove the unrendered template vars, but the bug I found with the missing images seems like it should be run down (even if conclusion is "we understand why this is happening and we can live with it).

If LOE on that is relatively high, might make sense to reduce this PR to just the change to solve for bug, and then handle the make command in a separate PR, in interest of getting ticket across board.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply README.md]: I added some more explanation in 162fd76c85216cc024f7ec854ba70af5fa5e2710. The problem (I think) is that Webpack injects script and style tags when serving up the SPA which don't get a nonce attribute applied to them. At least, I think that is what's going on.

File: revengine/urls.py (hunk 2 above)
Comment (nrh-cklimas) [reply revengine/urls.py:63]: Fixed this in dc4e6da270f60fb612a3a89fb3cf4e92ea02529d. The issue was that the proxy was trying to send images as text content.

File: revengine/views.py [hunk 3]
```diff
@@ -54,7 +58,39 @@ def _add_gtm_id(self, context):
...
+# Proxies the single page app in local development, parsing any HTML responses
+# using Django's template parser. If this is used, the single page app must have
+# its own dev proxying turned off for this to work. See spa/src/setupProxy.js
+# for how that's done.
+#
+# This was cribbed from
+# https://fractalideas.com/blog/making-react-and-django-play-well-together-hybrid-app-model/
+
+
+def proxy_spa_dev_server(request, upstream="http://localhost:3000"):
```
Comment (x110dc) [inline revengine/views.py:70]: clever

**Review by x110dc — CHANGES_REQUESTED** [review]

File: revengine/urls.py [hunk 4]
```diff
@@ -55,12 +55,9 @@
 
 if settings.DEBUG:  # pragma: no cover Is covered, but reimport confuses coverage
     from django.conf.urls.static import static
-    from django.contrib.staticfiles.urls import staticfiles_urlpatterns
 
     import debug_toolbar
```
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply revengine/urls.py:59]: My read of it is, git blame points to https://github.com/newsrevenuehub/rev-engine/commit/6756ba2b92d194f2bf5cff44895c0411013df88d where this was set up. It looks like at that point, the SPA was configured to build continuously to the Django static directory (e.g. look at webpack.local.config.js in that commit). At some point, our tooling drifted apart... create-react-app doesn't generate bundled files during local dev, so there are no files to serve up there anymore.

[the remark this answers is not in the document]
Comment (x110dc) [reply revengine/urls.py:59]: Got it. Thanks!

**Review by x110dc — APPROVED** [review]

**Review by benjaminEwhite — APPROVED** [review]

**Pull request merged**